File: library_qojqva/src/main/java/com/qojqva/impl/QojqvaFragment.java
package com.qojqva.impl;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.util.ArraySet;

import com.qojqva.entity.Permission;
import com.qojqva.internal.IPermissionCallback;
import com.qojqva.toolkit.PermissionKit;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

/**
 * @author #Suyghur.
 * Created on 3/29/2021
 */
@SuppressWarnings("deprecation")
@TargetApi(Build.VERSION_CODES.M)
public final class QojqvaFragment extends Fragment implements Runnable {

    private boolean mSpecialRequest = false;
    private boolean mDangerousRequestr = false;
    /**
     * 权限申请标记
     */
    public boolean mRequestFlag = false;
    public IPermissionCallback mCallback = null;
    private int mScreenOrientation = 0;

    /**
     * 请求的权限组
     */
    private static final String REQUEST_PERMISSIONS = "request_permissions";

    /**
     * 请求码（自动生成）
     */
    private static final String REQUEST_CODE = "request_code";

    /**
     * 权限请求码存放集合
     */
    private static final ArraySet<Integer> REQUEST_CODE_ARRAY = new ArraySet<>();


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Activity activity = getActivity();
        // 如果当前没有锁定屏幕方向就获取当前屏幕方向并进行锁定
        mScreenOrientation = activity.getRequestedOrientation();
        if (mScreenOrientation != ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED) {
            return;
        }
        int activityOrientation = getActivity().getResources().getConfiguration().orientation;
        try {
            // 兼容问题：在 Android 8.0 的手机上可以固定 Activity 的方向，但是这个 Activity 不能是透明的，否则就会抛出异常
            // 复现场景：只需要给 Activity 主题设置 <item name="android:windowIsTranslucent">true</item> 属性即可
            switch (activityOrientation) {
                case Configuration.ORIENTATION_LANDSCAPE:
                    getActivity().setRequestedOrientation(PermissionKit.isActivityReverse(activity) ? ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE : ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
                    break;
                case Configuration.ORIENTATION_PORTRAIT:
                default:
                    activity.setRequestedOrientation(PermissionKit.isActivityReverse(activity) ? ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT : ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                    break;
            }
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        // 如果当前 Fragment 是通过系统重启应用触发的，则不进行权限申请
        if (!mRequestFlag) {
            detachActivity(getActivity());
            return;
        }

        // 如果在 Activity 不可见的状态下添加 Fragment 并且去申请权限会导致授权对话框显示不出来
        // 所以必须要在 Fragment 的 onResume 来申请权限，这样就可以保证应用回到前台的时候才去申请权限
        if (mSpecialRequest) {
            return;
        }

        mSpecialRequest = true;
        requestSpecialPermission();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        Activity activity = getActivity();
        if (activity == null || mScreenOrientation != ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED) {
            return;
        }
        // 为什么这里不用跟上面一样 try catch ？因为这里是把 Activity 方向取消固定，只有设置横屏或竖屏的时候才可能触发 crash
        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mCallback = null;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        Bundle arguments = getArguments();
        FragmentActivity activity = getActivity();
        if (activity == null || arguments == null || mCallback == null || requestCode != arguments.getInt(REQUEST_CODE)) {
            return;
        }
        IPermissionCallback callback = mCallback;
        mCallback = null;

        // 优化权限回调结果
        PermissionKit.optimizePermissionResults(activity, permissions, grantResults);

        // 释放对这个请求码的占用
        REQUEST_CODE_ARRAY.remove(requestCode);
        // 将Fragment从Activity移除
        detachActivity(activity);

        // 获取已授予的权限
        List<String> grantedPermission = PermissionKit.getGrantedPermissions(permissions, grantResults);

        // 如果请求成功的权限集合大小和请求的数据一样大时证明权限已经全部授予
        if (grantedPermission.size() == permissions.length) {
            callback.onGranted(grantedPermission, true);
            return;
        }

        // 获取被拒绝的权限
        List<String> deiedPermission = PermissionKit.getGrantedPermissions(permissions, grantResults);

        callback.onDenied(deiedPermission, PermissionKit.isPermissionPermanentDenied(activity, deiedPermission));

        // 证明还有一部分权限被成功授予，回调成功接口
        if (!grantedPermission.isEmpty()) {
            callback.onGranted(grantedPermission, false);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        Bundle arguments = getArguments();
        FragmentActivity activity = getActivity();

        if (activity == null || arguments == null || mDangerousRequestr || requestCode != arguments.getInt(REQUEST_CODE)) {
            return;
        }

        mDangerousRequestr = true;
        // 需要延迟执行，不然有些华为机型授权了但是获取不到权限
        activity.getWindow().getDecorView().postDelayed(this, 300);
    }

    @Override
    public void run() {
        // 如果用户离开太久，会导致 Activity 被回收掉
        // 所以这里要判断当前 Fragment 是否有被添加到 Activity
        // 可在开发者模式中开启不保留活动来复现这个 Bug
        if (!isAdded()) {
            return;
        }
        // 请求其他危险权限
        requestDangerousPermission();
    }

    /**
     * 申请特殊权限
     */
    public void requestSpecialPermission() {
        Bundle arguments = getArguments();
        FragmentActivity activity = getActivity();
        if (arguments == null || activity == null) {
            return;
        }

        List<String> permissions = arguments.getStringArrayList(REQUEST_PERMISSIONS);

        // 是否需要申请特殊权限
        boolean requestSpecialPermission = false;

        // 判断当前是否包含特殊权限
        for (String permission : permissions) {
            if (PermissionKit.isSpecialPermission(permission)) {
                if (PermissionKit.isGrantedPermission(activity, permission)) {
                    // 已经授予过了，可以跳过
                    continue;
                }
                if (Permission.MANAGE_EXTERNAL_STORAGE.equals(permission) && !PermissionKit.isAndroid11()) {
                    // 当前必须是 Android 11 及以上版本，因为在旧版本上是拿旧权限做的判断
                    continue;
                }
                // 跳转到特殊权限授权页面
                startActivityForResult(PermissionSettingPage.getSmartPermissionIntent(activity, PermissionKit.asArrayList(permission)), getArguments().getInt(REQUEST_CODE));
                requestSpecialPermission = true;
            }
        }

        if (requestSpecialPermission) {
            return;
        }
        // 如果没有跳转到特殊权限授权页面，就直接申请危险权限
        requestDangerousPermission();
    }

    public void requestDangerousPermission() {
        Bundle arguments = getArguments();
        FragmentActivity activity = getActivity();
        if (activity == null || arguments == null) {
            return;
        }

        final int requestCode = arguments.getInt(REQUEST_CODE);

        final ArrayList<String> allPermissions = arguments.getStringArrayList(REQUEST_PERMISSIONS);
        if (allPermissions == null || allPermissions.size() == 0) {
            return;
        }

        ArrayList<String> locationPermission = null;
        // Android 10 定位策略发生改变，申请后台定位权限的前提是要有前台定位权限（授予了精确或者模糊任一权限）
        if (PermissionKit.isAndroid10() && allPermissions.contains(Permission.ACCESS_BACKGROUND_LOCATION)) {
            locationPermission = new ArrayList<>();
            if (allPermissions.contains(Permission.ACCESS_COARSE_LOCATION)) {
                locationPermission.add(Permission.ACCESS_COARSE_LOCATION);
            }

            if (allPermissions.contains(Permission.ACCESS_FINE_LOCATION)) {
                locationPermission.add(Permission.ACCESS_FINE_LOCATION);
            }
        }

        if (!PermissionKit.isAndroid10() || locationPermission == null || locationPermission.isEmpty()) {
            requestPermissions(allPermissions.toArray(new String[allPermissions.size() - 1]), getArguments().getInt(REQUEST_CODE));
            return;
        }

        // 在 Android 10 的机型上，需要先申请前台定位权限，再申请后台定位权限
        QojqvaFragment.beginRequest(activity, locationPermission, new IPermissionCallback() {

            @Override
            public void onGranted(List<String> permissions, boolean all) {
                if (!all || !isAdded()) {
                    return;
                }
                // 前台定位权限授予了，现在申请后台定位权限
                beginRequest(activity, PermissionKit.asArrayList(Permission.ACCESS_BACKGROUND_LOCATION), new IPermissionCallback() {
                    @Override
                    public void onGranted(List<String> permissions, boolean all) {
                        if (!all || !isAdded()) {
                            return;
                        }

                        // 前台定位权限和后台定位权限都授予了
                        int[] grantResults = new int[allPermissions.size()];
                        Arrays.fill(grantResults, PackageManager.PERMISSION_GRANTED);
                        onRequestPermissionsResult(requestCode, allPermissions.toArray(new String[0]), grantResults);
                    }

                    @Override
                    public void onDenied(List<String> permissions, boolean never) {
                        if (!isAdded()) {
                            return;
                        }

                        // 后台定位授权失败，但是前台定位权限已经授予了
                        int[] grantResults = new int[allPermissions.size()];
                        for (int i = 0; i < allPermissions.size(); i++) {
                            grantResults[i] = Permission.ACCESS_BACKGROUND_LOCATION.equals(allPermissions.get(i)) ? PackageManager.PERMISSION_DENIED : PackageManager.PERMISSION_GRANTED;
                        }
                        onRequestPermissionsResult(requestCode, allPermissions.toArray(new String[0]), grantResults);
                    }

                    @Override
                    public void onProxyFinish() {

                    }
                });
            }

            @Override
            public void onDenied(List<String> permissions, boolean never) {
                if (!isAdded()) {
                    return;
                }

                // 前台定位授权失败，并且无法申请后台定位权限
                int[] grantResults = new int[allPermissions.size()];
                Arrays.fill(grantResults, PackageManager.PERMISSION_DENIED);
                onRequestPermissionsResult(requestCode, allPermissions.toArray(new String[0]), grantResults);
            }

            @Override
            public void onProxyFinish() {

            }
        });
    }

    /**
     * 开启权限申请
     */
    public static void beginRequest(FragmentActivity activity, ArrayList<String> permissions, IPermissionCallback callback) {
        QojqvaFragment fragment = new QojqvaFragment();
        Bundle bundle = new Bundle();
        int requestCode;
        // 请求码随机生成，避免随机产生之前的请求码，必须进行循环判断
        do {
            // 新版本的 Support 库限制请求码必须小于 65536
            // 旧版本的 Support 库限制请求码必须小于 256
            requestCode = new Random().nextInt((int) Math.pow(2, 8));
        } while (REQUEST_CODE_ARRAY.contains(requestCode));
        // 标记这个请求码已经被占用
        REQUEST_CODE_ARRAY.add(requestCode);
        bundle.putInt(REQUEST_CODE, requestCode);
        bundle.putStringArrayList(REQUEST_PERMISSIONS, permissions);
        fragment.setArguments(bundle);
        // 设置保留实例，不会因为屏幕方向或配置变化而重新创建
        fragment.setRetainInstance(true);
        // 设置权限申请标记
        fragment.mRequestFlag = true;
        // 设置权限回调监听
        fragment.mCallback = callback;
        // 绑定到 Activity 上面
        fragment.attachActivity(activity);
    }


    /**
     * 绑定Activity
     */
    private void attachActivity(FragmentActivity activity) {
        activity.getSupportFragmentManager().beginTransaction().add(this, this.toString()).commitAllowingStateLoss();
    }

    /**
     * 解绑Activity
     */
    private void detachActivity(FragmentActivity activity) {
        activity.getSupportFragmentManager().beginTransaction().remove(this).commitAllowingStateLoss();
    }
}

File: library_qojqva/src/main/java/com/qojqva/impl/QojqvaFragment.kt
package com.qojqva.impl

import android.content.Context
import android.content.Intent
import android.content.pm.ActivityInfo
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.util.SparseBooleanArray
import com.qojqva.Qojqva

import com.qojqva.entity.Permission
import com.qojqva.internal.IPermissionCallback
import com.qojqva.toolkit.LogKit
import com.qojqva.toolkit.PermissionKit
import java.util.*

/**
 * @author #Suyghur.
 * Created on 3/29/2021
 */
class QojqvaFragment : Fragment(), Runnable {

    private var mSpecialRequest = false
    private var mDangerousRequest = false
    private var callback: IPermissionCallback? = null
    private var mScreenOrientation = 0

    override fun onAttach(context: Context) {
        super.onAttach(context)
        //如果当前没有锁定屏幕方向就获取当前屏幕方向并进行锁定
        mScreenOrientation = activity.requestedOrientation
        if (mScreenOrientation != ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED) {
            return
        }
        val activityOrientation = activity.resources.configuration.orientation
        try {
            //java.lang.IllegalStateException: Only fullscreen activities can request orientation
            //8.0系统BUG，在 Android 8.0 的手机上可以固定 Activity 的方向，但是这个 Activity 不能是透明的，否则就会抛出异常
            //只需要给 Activity 主题设置 <item name="android:windowIsTranslucent">true</item> 属性即可
            if (activityOrientation == Configuration.ORIENTATION_LANDSCAPE) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
            } else if (activityOrientation == Configuration.ORIENTATION_PORTRAIT) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
            }
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }
    }

    override fun onResume() {
        super.onResume()
        //如果在 Activity 不可见的状态下添加 Fragment 并且去申请权限会导致授权对话框显示不出来
        //所以必须要在 Fragment 的 onResume 来申请权限，这样就可以保证应用回到前台的时候才去申请权限
        if (mSpecialRequest) {
            return
        }
        mSpecialRequest = true
        requestSpecialPermission()
    }

    override fun onDestroy() {
        super.onDestroy()
        callback = null
    }

    override fun onDetach() {
        super.onDetach()
        if (mScreenOrientation == ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED) {
            return
        }
        //取消方向固定
        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (callback == null || requestCode != arguments.getInt(REQUEST_CODE)) {
            return
        }
        val tmpCallback = callback!!
        for (i in permissions.indices) {
            val permission = permissions[i]
            if (PermissionKit.isSpecialPermission(permission)) {
                LogKit.d("这个权限是特殊权限，重新开始检查")
                //如果这个权限是特殊权限，那么就重新开始检查
                grantResults[i] = PermissionKit.getPermissionStatus(activity, permission)
                continue
            }

            //重新检查11.0后台定位权限
            if (PermissionKit.isAndroid11 && Permission.ACCESS_BACKGROUND_LOCATION == permission) {
                LogKit.d("重新检查11.0后台定位权限")
                //这个权限是后台定位权限并且当前手机版本是 Android 11 及以上，那么就需要重新进行检查
                //因为只要申请这个后台定位权限，grantResults 数组总对这个权限申请的结果返回 -1（拒绝）
                grantResults[i] = PermissionKit.getPermissionStatus(activity, permission)
                continue
            }

            //重新检查10.0的三个新权限
            if (!PermissionKit.isAndroid10 && (Permission.ACCESS_BACKGROUND_LOCATION == permission
                            || Permission.ACTIVITY_RECOGNITION == permission) || Permission.ACCESS_MEDIA_LOCATION == permission) {
                LogKit.d("重新检查10.0的三个新权限")
                //如果当前版本不符合最低要求，那么就重新进行检查
                grantResults[i] = PermissionKit.getPermissionStatus(activity, permission)
                continue
            }

            //重新检查9.0的新权限
            if (!PermissionKit.isAndroid9 && Permission.ACCEPT_HANDOVER == permission) {
                //如果当前版本不符合最低要求，那么就重新进行检查
                LogKit.d("重新检查9.0的新权限")
                grantResults[i] = PermissionKit.getPermissionStatus(activity, permission)
                continue
            }

            //重新检查8.0的两个新权限
            if (!PermissionKit.isAndroid8 && (Permission.ANSWER_PHONE_CALLS == permission || Permission.READ_PHONE_NUMBERS == permission)) {
                //如果当前版本不符合最低要求，那么就重新进行检查
                LogKit.d("重新检查8.0的两个新权限")
                grantResults[i] = PermissionKit.getPermissionStatus(activity, permission)
            }
        }

        //释放请求码
        requestCodeArray.delete(requestCode)
        //将 Fragment 从 Activity 移除
        detachActivity(activity)

        //获取已授予的权限
        val grantedPermissions = PermissionKit.getGrantedPermissions(permissions, grantResults)
        //如果请求成功的权限集合大小和请求的数组一致则证明全部授予
        if (grantedPermissions.size == permissions.size) {
            Qojqva.interceptor.grantedPermissions(activity, grantedPermissions, true, tmpCallback)
            return
        }

        //获取被拒绝的权限
        val deniedPermissions = PermissionKit.getDeniedPermissions(permissions, grantResults)
        //申请的权限有不同意的，如果有个权限被永久拒绝就返回true，让用户跳转到设置界面开启
        LogKit.d("申请的权限有不同意的，如果有个权限被永久拒绝就返回true，让用户跳转到设置界面开启")
        Qojqva.interceptor.deniedPermissions(activity, deniedPermissions, PermissionKit.isPermissionPermanentDenied(activity, deniedPermissions), tmpCallback)

        if (grantedPermissions.isNotEmpty()) {
            //证明还有一部分权限被成功授予，回调成功接口
            Qojqva.interceptor.grantedPermissions(activity, grantedPermissions, false, tmpCallback)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode != arguments.getInt(REQUEST_CODE) || mDangerousRequest) {
            return
        }
        mDangerousRequest = true
        //需要延迟执行，不然有些华为机型授权了但是没效果
        activity.window.decorView.postDelayed(this, 200)
    }

    override fun run() {
        //如果用户离开太久，会导致 Activity 被回收掉
        //所以这里要判断当前 Fragment 是否有被添加到 Activity
        //可在开发者模式中开启不保留活动来复现这个 Bug
        if (!isAdded) {
            return
        }
        //请求其他危险权限
        requestDangerousPermission()
    }

    /**
     * 绑定Activity
     */
    private fun attachActivity(activity: FragmentActivity) {
        activity.supportFragmentManager.beginTransaction().add(this, this.toString()).commitAllowingStateLoss()
    }

    /**
     * 解绑Activity
     */
    private fun detachActivity(activity: FragmentActivity) {
        activity.supportFragmentManager.beginTransaction().remove(this).commitAllowingStateLoss()
    }

    private fun requestSpecialPermission() {
        val permissions = arguments.getStringArrayList(REQUEST_PERMISSIONS)
        if (permissions.isNullOrEmpty()) {
            //如果空则直接返回
            return
        }
        //是否需要申请特殊权限
        var requestSpecialPermission = false
        //判断当前是否包含特殊权限
        if (PermissionKit.containsSpecialPermission(permissions)) {
            if (permissions.contains(Permission.MANAGE_EXTERNAL_STORAGE) && !PermissionKit.isGrantedSettingPermission(activity)) {
                //当前必须是 Android 11 及以上版本，因为 hasStoragePermission 在旧版本上是拿旧权限做的判断，所以这里需要多判断一次版本
                if (PermissionKit.isAndroid11) {
                    //跳转到存储权限设置界面
                    startActivityForResult(PermissionSettingPage.getStoragePermissionIntent(activity), arguments.getInt(REQUEST_CODE))
                    requestSpecialPermission = true
                }
            }

            if (permissions.contains(Permission.REQUEST_INSTALL_PACKAGES) && !PermissionKit.isGrantedInstallPermission(activity)) {
                //调整到安装权限设置界面
                startActivityForResult(PermissionSettingPage.getInstallPermissionIntent(activity), arguments.getInt(REQUEST_CODE))
                requestSpecialPermission = true
            }

            if (permissions.contains(Permission.SYSTEM_ALERT_WINDOW) && !PermissionKit.isGrantedWindowPermission(activity)) {
                //跳转到悬浮窗设置界面
                startActivityForResult(PermissionSettingPage.getWindowPermissionIntent(activity), arguments.getInt(REQUEST_CODE))
                requestSpecialPermission = true
            }

            if (permissions.contains(Permission.NOTIFICATION_SERVICE) && !PermissionKit.isGrantedNotifyPermission(activity)) {
                // 跳转到通知栏权限设置页面
                startActivityForResult(PermissionSettingPage.getNotifyPermissionIntent(activity), arguments.getInt(REQUEST_CODE))
                requestSpecialPermission = true
            }

            if (permissions.contains(Permission.WRITE_SETTINGS) && !PermissionKit.isGrantedSettingPermission(activity)) {
                // 跳转到系统设置权限设置页面
                startActivityForResult(PermissionSettingPage.getSettingPermissionIntent(activity), arguments.getInt(REQUEST_CODE))
                requestSpecialPermission = true
            }
        }
        //当前必须没有跳转到悬浮窗或安装权限界面
        if (!requestSpecialPermission) {
            requestDangerousPermission()
        }
    }

    /**
     * 申请危险权限
     */
    private fun requestDangerousPermission() {
        val allPermissions = arguments.getStringArrayList(REQUEST_PERMISSIONS)
        if (allPermissions.isNullOrEmpty()) {
            //如果空则直接返回
            return
        }
        val locationPermissions = arrayListOf<String>()
        //Android 10 定位策略发生改变，申请后台定位权限的前提是要有前台定位权限（授予了精确或者模糊任一权限）
        if (PermissionKit.isAndroid10 && allPermissions.contains(Permission.ACCESS_BACKGROUND_LOCATION)) {
            if (allPermissions.contains(Permission.ACCESS_COARSE_LOCATION) && !PermissionKit.isGrantedPermission(activity, Permission.ACCESS_COARSE_LOCATION)) {
                locationPermissions.add(Permission.ACCESS_COARSE_LOCATION)
            }
            if (allPermissions.contains(Permission.ACCESS_FINE_LOCATION) && !PermissionKit.isGrantedPermission(activity, Permission.ACCESS_FINE_LOCATION)) {
                locationPermissions.add(Permission.ACCESS_FINE_LOCATION)
            }
        }

        //如果不需要申请前台定位权限就直接申请危险权限
        if (locationPermissions.isNullOrEmpty()) {
            requestPermissions(allPermissions.toArray(arrayOfNulls(allPermissions.size - 1)), arguments.getInt(REQUEST_CODE))
            return
        }

        //在 Android 10 的机型上，需要先申请前台定位权限，再申请后台定位权限
        beginRequest(activity, locationPermissions, object : IPermissionCallback {
            override fun onGranted(permissions: ArrayList<String>, all: Boolean) {
                if (!all || !isAdded) {
                    return
                }
                requestPermissions(allPermissions.toArray(arrayOfNulls(allPermissions.size - 1)), arguments.getInt(REQUEST_CODE))
            }

            override fun onDenied(permissions: ArrayList<String>, never: Boolean) {
                if (!isAdded) {
                    return
                }
                //如果申请的权限里面只包含定位相关权限，那么就直接返回失败
                if (permissions.size == allPermissions.size - 1) {
                    val grantResults = IntArray(allPermissions.size) { 0 }
                    Arrays.fill(grantResults, PackageManager.PERMISSION_DENIED)
                    onRequestPermissionsResult(arguments.getInt(REQUEST_CODE), allPermissions.toTypedArray(), grantResults)
                    return
                }
                //如果还有其他类型的权限组就继续申请
                requestPermissions(allPermissions.toArray(arrayOfNulls(allPermissions.size - 1)), arguments.getInt(REQUEST_CODE))
            }

            override fun onProxyFinish() {
            }
        })
    }


    companion object {
        /**
         * 请求的权限组
         */
        private const val REQUEST_PERMISSIONS = "request_permissions"

        /**
         * 请求码（自动生成）
         */
        private const val REQUEST_CODE = "request_code"

        /**
         * 权限请求码存放集合
         */
        private val requestCodeArray = SparseBooleanArray()

        /**
         * 开启权限申请
         */
        fun beginRequest(activity: FragmentActivity, permissions: ArrayList<String>, callback: IPermissionCallback) {
            val fragment = QojqvaFragment()
            val bundle = Bundle()
            var requestCode: Int
            //请求码随机生成，避免随机产生之前的请求码，必须进行循环判断
            do {
                requestCode = PermissionKit.getRandomRequestCode()
            } while (requestCodeArray.get(requestCode))
            //标记这个请求码已被占用
            requestCodeArray.put(requestCode, true)
            bundle.putInt(REQUEST_CODE, requestCode)
            bundle.putStringArrayList(REQUEST_PERMISSIONS, permissions)
            fragment.arguments = bundle
            //设置保存实例，不会因为屏幕方向或配置变化重新创建
            fragment.retainInstance = true
            //设置权限回调
            fragment.callback = callback
            //绑定到Activity上
            fragment.attachActivity(activity)
        }
    }
}